### PR TITLE
Send error log to Discord

### PR DESCRIPTION
**Pull request by tayler6000**

### CODE REVIEW

File: servicemodules/discord.py [hunk 1]
```diff
@@ -99,7 +102,16 @@ async def on_message(message: discord.message.Message) -> None:
             responses = [Response()]
             for module in self.modules:
                 log.info(class_name, msg=f"# Asking module: {module}")
-                response = module.process_message(message)
+                try:
+                    response = module.process_message(message)
+                except Exception as e:
```
Comment (MatthewBaggins) [inline servicemodules/discord.py:107]: Can you use a more narrow exception? What exact kind of error does module.process_message throw that you want to catch?

Comment (tayler6000) [reply servicemodules/discord.py:107]: They're not _supposed_ to throw anything except maybe `SystemExit` in the case of `s, reboot` that's why it needs to catch everything under `Exception` to report any bugs to the bot devs.  The only other alternative is to have Rob SSH into the bot to check logs, which isn't ideal.  Without this, there is a `try: except Exception:` wrapping `on_message` in the discord py backend dropping the error.  This try catch makes the error handling actually useful, and allows the message to be responded to even if one module fails. (Currently any failure prevents a response)

**Review by Aprillion — APPROVED** [review]

**Comment by robertskmiles** [comment]
How do we test this? Is there a know exception-throwing bug we can use?

**Comment by Aprillion** [comment]
@robertskmiles  do you prefer to NOT merge it until we experience a bug where we will need the logs?

or merge it now to see how it helps with the future unexpected bug?

or include a module that would raise an exception for `stampy, show me how exceptional you are!` prompt?

**Comment by robertskmiles** [comment]
@Aprillion That last one is hilarious! Maybe just as part of the stampycontrols module? But if not that I'd want at least someone to run a version with a deliberate crash on the test server (I'm assuming that hasn't happened because the `#stampy-test-log` channel on the test server is empty)

**Review by Aprillion — CHANGES_REQUESTED** [review]
huh, indeed this doesn't work for me on test server, trying to debug right now

File: servicemodules/discord.py [hunk 2]
```diff
@@ -99,7 +102,16 @@ async def on_message(message: discord.message.Message) -> None:
...
+                    parts = ["Traceback (most recent call last):\n"]
+                    parts.extend(traceback.format_stack(limit=25)[:-2])
+                    parts.extend(traceback.format_exception(*sys.exc_info())[1:])
+                    message = "".join(parts)
+                    if self.error_channel is None:
+                        self.error_channel = await client.get_channel(error_channel_id)
```
Comment (Aprillion) [inline servicemodules/discord.py:113]: `client` is not defined here

**Comment by Aprillion** [comment]
### post mortem discord discussion
* we had a debugging session with TJ on stampy's test server discord (https://discord.com/channels/783123903382814720/803448149946662923/1020369024900665475 and https://discord.com/channels/783123903382814720/1017531179664150608/1020377195073515550)
* the core problem(s) turned out to undefined `client` and and extra `await` for a non-async function in this PR
* when fixed, we discovered a few modules where already throwing exception to almost every message right now `AttributeError: 'str' object has no attribute 'clean_content'`
* to avoid spamming Discord (risking stampy to get banned because of DDoS), TJ is going to fix the troublesome modules in this PR - after getting home to charge his dying laptop - some problems with the superclass and what not (to be seen later in this PR)

**Comment by tayler6000** [comment]
Turns out these errors were caused by the PR itself, just a variable name collision.  I didn't have the keys to the Stampy Test Server until today so this had gone untested before today.

PR should be good to merge now.

**Review by Aprillion — APPROVED** [review]
LGTM now on test server

**Comment by tayler6000** [comment]
@robertskmiles Ok to merge?

**Pull request merged**